### PR TITLE
Provide metrics APIs for agent plugin

### LINKED ISSUE #5494: Support metrics APIs for agent plugin
In the 8.1.0, we have supported our own toolkit library and MicroMeter integration for metrics in the java application. Now, ShardingSphere community is requesting to collect metrics in the agent plugin system.

Right now, the `transform` and `adapter` systems are for toolkit with activation. In the agent core, we need this kind of simple APIs, such as `MeterFactory#createCounter`, for use inside the agent.

### PR DESCRIPTION
**Pull request by mrproliu**
2. Move toolkit counter rate mode to the core level

Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues
Resolved #5494 

### New feature or improvement
- Provide metrics APIs for agent plugin
- Move toolkit counter rate mode to the core level

### CODE REVIEW

File: docs/en/guides/Java-Plugin-Development-Guide.md [hunk 1]
```diff
@@ -443,3 +443,44 @@ Please follow there steps:
...
+import org.apache.skywalking.apm.agent.core.meter.MeterFactory;
+
+Counter counter = MeterFactory.counter(meterName).tag("tagKey", "tagValue").mode(Counter.Mode.INCREMENT).build();
+counter.increment(1d);
+```
+1. `MeterFactory.counter` Create a new counter builder with the meter name.
+1. `Counter.Builder.tag(String key, String value)` Mark a tag key/value pair.
+1. `Counter.Builder.mode(Counter.Mode mode)` Change the counter mode, `RATE` mode means reporting rate to the backend.
+1. `Counter.Builder.build()` Build a new `Counter` which is collected and reported to the backend.
+1. `Counter.increment(double count)` Increment count to the `Counter`, It could be a positive/negative value.
```
Comment (wu-sheng) [inline docs/en/guides/Java-Plugin-Development-Guide.md:461]: I feel accepting the `negative` value is conflicting with the definition of the counter. `a single monotonically increasing counter`. 

Please recheck.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/meter/builder/Histogram.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.agent.core.meter.builder;
+
+import java.util.List;
+
+/**
+ * Similar to a histogram, a summary sample observations (usual things like request durations and response sizes).
```
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/meter/builder/Histogram.java:24]: What does this `similar to a histogram` mean? Isn't this a histogram already?

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/meter/builder/Histogram.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+ * While it also provides a total count of observations and a sum of all observed values, it calculates configurable quartiles over a sliding time window.
```
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/meter/builder/Histogram.java:25]: Are you copying this from Prometheus or somewhere? Also, at the same time, `HistogramImpl`'s comments don't make sense.

File: apm-application-toolkit/apm-toolkit-meter/src/main/java/org/apache/skywalking/apm/toolkit/meter/impl/CounterImpl.java [hunk 4]
```diff
@@ -51,30 +48,13 @@ public void increment(double count) {
     /**
      * Get count value
      */
-    public double get() {
+    public double getCount() {
```
Comment (wu-sheng) [inline apm-application-toolkit/apm-toolkit-meter/src/main/java/org/apache/skywalking/apm/toolkit/meter/impl/CounterImpl.java:51]: Why change this name? But such as in `GuageImpl`, still, use `get`? Seems inconsistent to me.

**Comment by wu-sheng** [comment]
Also, the CI seems to fail somehow, please recheck.

File: apm-application-toolkit/apm-toolkit-meter/src/main/java/org/apache/skywalking/apm/toolkit/meter/impl/CounterImpl.java (hunk 4 above)
Comment (mrproliu) [reply apm-application-toolkit/apm-toolkit-meter/src/main/java/org/apache/skywalking/apm/toolkit/meter/impl/CounterImpl.java:51]: Change them to the same function name, using `getCount` to get meter value. I forget to change the `Gauge`, has resolved.

File: docs/en/guides/Java-Plugin-Development-Guide.md [hunk 5]
```diff
@@ -443,3 +443,44 @@ Please follow there steps:
 1. Send the pull request and ask for review. 
 1. The plugin committers approve your plugins, plugin CI-with-IT, e2e and plugin tests passed.
 1. The plugin accepted by SkyWalking. 
+
+### Plugin Meter reporter
```
Comment (wu-sheng) [inline docs/en/guides/Java-Plugin-Development-Guide.md:447]:
```suggestion
### Plugin Meter Plugin
```

File: docs/en/guides/Java-Plugin-Development-Guide.md [hunk 6]
```diff
@@ -443,3 +443,44 @@ Please follow there steps:
...
+Java agent also supports customize meter report to backend. We provide Java Agent Core level API such as [Application Toolkit meter](../setup/service-agent/java-agent/Application-toolkit-meter.md).
```
Comment (wu-sheng) [inline docs/en/guides/Java-Plugin-Development-Guide.md:448]:
```suggestion
Java agent plugin could use meter APIs to collect the metrics for backend analysis.
```

**Review by wu-sheng — APPROVED** [review]

**Review by wu-sheng — CHANGES_REQUESTED** [review]
e2e of meter system fails.

**Comment by yu199195** [comment]
@mrproliu  Not yet, it seems.

**Comment by wu-sheng** [comment]
@yu199195 This PR is not ready to be merged, there are 2 things left
1. The implementations of meter core are duplicated in the toolkit and agent core both. I have informed @mrproliu, he need to merge them and re-implement by referring the OpenTracing toolkit way. Toolkit lib should include a shell only.
1. After (1), we need to deploy the snapshot version for the API toolkit on the Apache snapshot repo, then the tests could pass, and you could use that toolkit.

**Comment by yu199195** [comment]
> @yu199195 This PR is not ready to be merged, there are 2 things left
>1.The implementations of meter core are duplicated in the toolkit and agent core both. I have informed @mrproliu, he need to >merge them and re-implement by referring the OpenTracing toolkit way. Toolkit lib should include a shell only.
>2.After (1), we need to deploy the snapshot version for the API toolkit on the Apache snapshot repo, then the tests could pass, and you could use that toolkit.

thanks, i will follow this pr

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**